### PR TITLE
feat: add Palette component

### PR DESCRIPTION
**Pull request by Ruminat**
DATAUI-2067

`Palette` — a basic grid with elements you can select/unselect (a grid of checkboxes basically).

image

Can be customized to display anything basically icons/GIFs/images/reactions/colors and so on:

image

### CODE REVIEW

**Comment by korvin89** [comment]
Hey @Ruminat! We've discussed ur pr with @amje @ValeraS and have some suggestions for u:
1. Add separate component `Palette` with this interface:
```typescript
interface PaletteOption {
    value: string;
    content: React.ReactNode;
    title?: string; // HTML title
    disabled?: boolean;
}

interface PaletteProps extends DOMProps, QAProps {
    value?: string[];
    defaultValue?: string[];
    onUpdate?: (value: string[]) => void;
    onFocus?: React.FocusEventHandler;
    onBlur?: React.FocusEventHandler;
    disabled?: boolean;
    name?: string;
    options: PaletteOption[];
    size?: ButtonSize;
}
```
`Palette` should have fully keyboard support:
- the first press of the tab key should focus first option in the component
- the second press of the tab key should take the focus away from the component
- navigation between options should be managed by arrow keys
2. Use `Palette` as a base for `EmojiPalette` 
3. Use [this](https://github.com/gravity-ui/uikit/tree/main/src/hooks/useControlledState) hook for value management

**Comment by Ruminat** [comment]
> Add separate component `Palette`

I'll add the `Palette` component (next week, I think)

Do we need the `EmojiPalette` component in `uikit` then?
It seems that we can simply put a usage example of `Palette` for `EmojiPalette` in storybook
But is there a need to make a distinct `EmojiPalette` component?

**Comment by korvin89** [comment]
> Do we need the `EmojiPalette` component in `uikit` then? It seems that we can simply put a usage example of `Palette` for `EmojiPalette` in storybook But is there a need to make a distinct `EmojiPalette` component?

Hmm, we probably don't need it. Example in Storybook sounds quite enough

**Comment by Ruminat** [comment]
Turned the component into `Palette`

**Comment by Ruminat** [comment]
@korvin89 @amje @ValeraS ping

**Comment by Ruminat** [comment]
@korvin89 @amje @ValeraS ping

File: src/components/Palette/Palette.tsx
Comment (korvin89) [inline src/components/Palette/Palette.tsx:231]: Suggestions about `PaletteOptionItem` component:
- rename ~~PaletteOptionItem~~ -> `PaletteOption`
- use `Button` as wrapper for option (you're already doing it) and add property `renderOption` (like in `Select`) and use `Button.Icon` by default. In this case, we will not need the following options: `optionClassName`, `iconClassName`
- rename `checked` -> selected

File: src/components/Palette/Palette.tsx
Comment (korvin89) [inline src/components/Palette/Palette.tsx:268]: I think we do not need `PaletteControl` at all. Looks like redundant abstraction

File: src/components/Palette/Palette.tsx
Comment (korvin89) [inline src/components/Palette/Palette.tsx:171]: There is no need in brackets in case of constants usage

File: src/components/Palette/Palette.tsx
Comment (korvin89) [inline src/components/Palette/Palette.tsx:184]: The same issue about constants

File: src/components/Palette/Palette.tsx
Comment (korvin89) [inline src/components/Palette/Palette.tsx:169]: You can check w3c pattern [here](https://www.w3.org/WAI/ARIA/apg/patterns/grid/) to clarify the roles of the elements

File: src/components/Palette/README.md
Comment (korvin89) [inline src/components/Palette/README.md:205]: Why do we need a description of the internal interface?

**Review by korvin89 — CHANGES_REQUESTED** [review]

File: src/components/Palette/Palette.tsx
Comment (Ruminat) [reply src/components/Palette/Palette.tsx]: Fixed

File: src/components/Palette/Palette.tsx
Comment (Ruminat) [reply src/components/Palette/Palette.tsx]: You're right
It's much simpler without it
Removed it

File: src/components/Palette/PaletteControl.tsx
Comment (Ruminat) [reply src/components/Palette/PaletteControl.tsx]: Removed the `checkbox` role and `aria-checked` attribute

File: src/components/Palette/Palette.tsx
Comment (Ruminat) [reply src/components/Palette/Palette.tsx]: Done

File: src/components/Palette/Palette.tsx
Comment (Ruminat) [reply src/components/Palette/Palette.tsx]: Removed the global listeners and used the `useFocusWithin` hook

File: src/components/Palette/Palette.tsx
Comment (Ruminat) [reply src/components/Palette/Palette.tsx:32]: Added jsdoc comments to all properties

File: src/components/Palette/Palette.tsx
Comment (Ruminat) [reply src/components/Palette/Palette.tsx:170]: I think I used the right roles except for the `checkbox` for the `Button`
I also added a wrapper for the `Button` with `gridcell` role (I'm not sure if it's necessary, but it's recommended in the article you sent if I understand correctly)
The `grid` and `row` roles seem to be right

**Comment by korvin89** [comment]
> @korvin89 ping

@Ruminat  Seems pretty good! Lets wait for at least another one review from @ValeraS or @amje

**Comment by Ruminat** [comment]
@amje ping

**Review by korvin89 — APPROVED** [review]

File: src/components/Palette/utils.ts
Comment (amje) [inline src/components/Palette/utils.ts:3]: Let's add some basic tests for this function

File: src/components/Palette/__stories__/Palette.stories.tsx
Comment (amje) [inline src/components/Palette/__stories__/Palette.stories.tsx:40]: Please use CSF3 format of stories, see an example in Button's stories

File: src/components/Palette/Palette.tsx
Comment (amje) [inline src/components/Palette/Palette.tsx:92]: Why not "m"? It's a default for most of our components

File: src/components/Palette/__stories__/Palette.stories.tsx
Comment (amje) [inline src/components/Palette/__stories__/Palette.stories.tsx:3]: Let's import only a few for example, not need to import all

File: src/components/Palette/Palette.scss
Comment (amje) [inline src/components/Palette/Palette.scss:33]: We need shift the sizes down, they're a bit bigger than they should be. You can check font sizes of Button

File: src/components/Palette/Palette.tsx
Comment (amje) [inline src/components/Palette/Palette.tsx:72]: `icon` name is misleading, icon is a way we center content in the Button. And why we need this class in the first place?

**Review by amje — CHANGES_REQUESTED** [review]

**Comment by amje** [comment]
In RTL mode keyboard navigation is broken

**Comment by Ruminat** [comment]
> In RTL mode keyboard navigation is broken

Fixed

File: src/components/Palette/Palette.tsx
Comment (Ruminat) [reply src/components/Palette/Palette.tsx]: I don't think we really need it then
Removed it

File: src/components/Palette/Palette.scss
Comment (Ruminat) [reply src/components/Palette/Palette.scss]: Shifted the font sizes down according to the `Button`'s `--_--icon-size`: `12px`/`16px`/`16px`/`16px`/`20px`
The font sizes of `Button` are basically the same (except for the `xl`), so I didn't use them.

File: src/components/Palette/__stories__/Palette.stories.tsx
Comment (Ruminat) [reply src/components/Palette/__stories__/Palette.stories.tsx]: Done

File: src/components/Palette/README.md
Comment (Ruminat) [reply src/components/Palette/README.md]: Added

File: src/components/Palette/__stories__/Palette.stories.tsx
Comment (Ruminat) [reply src/components/Palette/__stories__/Palette.stories.tsx]: Done

File: src/components/Palette/utils.ts
Comment (Ruminat) [reply src/components/Palette/utils.ts:3]: Added tests

**Comment by Ruminat** [comment]
@amje ping

**Review by amje — APPROVED** [review]

**Pull request merged**
